liesel/goose/optim.py

```diff
@@ -0,0 +1,551 @@
+
+    Next, generate some data.
+
+    >>> key = jax.random.PRNGKey(42)
+    >>> key, subkey = jax.random.split(key)
+    >>> x = jax.random.normal(key, (100,))
+    >>> y = 0.5 + 1.2 * x + jax.random.normal(subkey, (100,))
+
+    Next, set up a linear model. For simplicity, we assume the scale to be fixed to the
+    true value of 1.
+
+    >>> coef = lsl.param(jnp.zeros(2), name="coef")
+    >>> xvar = lsl.obs(jnp.c_[jnp.ones_like(x), x], name="x")
+    >>> mu = lsl.Var(lsl.Calc(jnp.dot, xvar, coef), name="mu")
+    >>> ydist = lsl.Dist(tfd.Normal, loc=mu, scale=1.0)
+    >>> yvar = lsl.obs(y, ydist, name="y")
+    >>> model = lsl.GraphBuilder().add(yvar).build_model()
+
+    Now, we are ready to run the optimization.
+
+    >>> result = gs.optim_flat(model, params=["coef"])
+    >>> {name: jnp.round(value, 2) for name, value in result.position.items()}
+    {'coef': Array([0.52, 1.29], dtype=float32)}
+
+    We can now, for example, use ``result.model_state`` in
+    :meth:`.EngineBuilder.set_initial_values` to implement a "warm start" of MCMC
+    sampling.
+
+    """
+    # ---------------------------------------------------------------------------------
+    # Validation input
+    if restore_best_position:
+        assert (
+            save_position_history
+        ), "Cannot restore best position if history is not saved."
+
+    # ---------------------------------------------------------------------------------
+    # Pre-process inputs
+
+    batch_seed = (
+        batch_seed if batch_seed is not None else np.random.randint(low=1, high=1000)
+    )
+
+    user_patience = stopper.patience
+    if model_validation is None:
+        model_validation = model_train
+        stopper.patience = stopper.max_iter
+
+    if optimizer is None:
+        optimizer = optax.adam(learning_rate=1e-2)
+
+    n_train = _find_sample_size(model_train)
+    n_validation = _find_sample_size(model_validation)
+    observed = _find_observed(model_train)
+
+    batch_size = batch_size if batch_size is not None else n_train
+
+    interface_train = LieselInterface(model_train)
+    position = interface_train.extract_position(params, model_train.state)
+    interface_train._model.auto_update = False
+
+    interface_validation = LieselInterface(model_validation)
+    interface_validation._model.auto_update = False
+
+    # ---------------------------------------------------------------------------------
+    # Validate model log prob decomposition
+    _validate_log_prob_decomposition(
+        interface_train, position=position, state=model_train.state
+    )
+    _validate_log_prob_decomposition(
+        interface_validation, position=position, state=model_validation.state
+    )
+
+    # ---------------------------------------------------------------------------------
+    # Define loss function(s)
+
+    likelihood_scalar = n_train / batch_size
+    likelihood_scalar_validation = n_train / n_validation
```

Comment:
I'm just curious. Why is that?

Reply:
Thanks for the question @GianmarcoCallegher !
Ah the idea here is to rescale the likelihood to keep its ratio to the prior constant, as discussed here: https://github.com/liesel-devs/liesel/pull/175#issuecomment-2030409019

Otherwise, if you use batching or a small test dataset, the prior gets relatively more important than if you were to use the likelihood. 

Does that make sense in your opinion?